[Tan Chin Kiat] iP

tanchinkiat99 opened the pull request:
### Mr Tutu
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/#11_Your_mind_is_for_having_ideas_not_holding_them_-_David_Allen))
Mr Tutu helps you remember all your tasks so you don't have to. It is

- easy to use
- quick to download
- and ~~$29.99/mth~~ _completely_ **FREE**

All you need to do is:

1. download it [here](https://github.com/nus-cs2103-AY2122S2/ip/pull/205),
2. add your tasks,
3. and let it manage your tasks for you! 😃 

Features:

- [x] Managing tasks
- [ ]  Managing deadlines (coming soon!)
- [ ] Reminders

If you Java programmer, you can use it to practice Java too. Here's a snippet of the `main` method:

``` ruby
public static void main(String[] args) throws InvalidInputException, IOException {
        Storage store = new Storage("./data");
        File f = store.initialize();

        Tutu.separator();
        System.out.println("Hello! I'm Tutu\nWhat can I do for you?");
        Tutu.separator();

        Ui ui = new Ui();
        String cmd = ui.read();
        while (!cmd.equals("bye")) {
            ...
        }
        System.out.println("Bye. Hope to see you again soon!");
    }
```

Commit:
Add Gradle support

tanchinkiat99 committed:
skeletal version of Duke (customized as Tutu)

tanchinkiat99 committed:
customized display format

tanchinkiat99 committed:
added add and list functionality

tanchinkiat99 committed:
added mark and unmark

tanchinkiat99 committed:
added functionality for event, deadline, and todo

tanchinkiat99 committed:
conducted i/o redirection test

tanchinkiat99 committed:
created and integrated exception class

tanchinkiat99 committed:
added delete functionality

tanchinkiat99 committed:
Tutu class: enabled creating of data file and saving of task list in the file

tanchinkiat99 committed:
Deadline class: enabled parsing of date and time input in format of yyyy-mm-dd HHmm
Event class: enabled parsing of date and time input in format of yyyy-mm-dd HHmm to HHmm

tanchinkiat99 committed:
Merge branch 'branch-Level-8'

tanchinkiat99 committed:
Ui class: extracted Ui class for user interaction functions including taking user input and loading error messages

tanchinkiat99 committed:
Storage class: extracted Storage class for creating data file, loading tasklist and saving tasks into file

tanchinkiat99 committed:
Parser class: for parsing the user's input command

tanchinkiat99 committed:
TaskList class: for containing the task list, and also its related functions like adding, deleting, marking and unmarking tasks

tanchinkiat99 committed:
Organized classes into packages

tanchinkiat99 committed:
Reorganized classes into packages, and added JUnit integration

tanchinkiat99 committed:
Added JUnit test to test some functionalities including adding, removing, marking and unmarking of Tasks of the TaskList class

tanchinkiat99 committed:
Added javadocs comments to all function bodies and classes.

tanchinkiat99 committed:
Edited code to follow basic and intermediate rules in the SE-EDU Java Coding Standard

tanchinkiat99 committed:
Add find functionality

tanchinkiat99 committed:
Remove redundant import

tanchinkiat99 committed:
Merge branch 'branch-A-CodingStandard'

tanchinkiat99 committed:
Resolved Merge Conflicts

tanchinkiat99 committed:
added find

tanchinkiat99 committed:
Merge branch 'master' into branch-Level-9

tanchinkiat99 committed:
Added documentation for find functionality.

atmh commented on src/main/java/tutu/Tutu.java:
```diff
@@ -0,0 +1,90 @@
+import tutu.data.Storage;
+import tutu.data.TaskList;
+import tutu.exception.InvalidInputException;
+import tutu.task.*;
```
Wildcard import should not be used. To stop IntelliJ from doing it automatically, you could change the settings here
`File > Settings > Editor > Code Style > Java > Class count to use import with '*':`

atmh commented on src/main/java/tutu/data/Storage.java:
```diff
@@ -0,0 +1,56 @@
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Scanner;
+
+public class Storage {
```
I like that you have included Javadocs for most of your classes and methods 😄 . 
Just a few more Javadocs to include in Storage.java, Parser.java and UI.java 👍.

atmh commented on src/main/java/tutu/user/Parser.java:
```diff
@@ -0,0 +1,32 @@
+
+public class Parser {
+
+    public enum Function {
+        LIST, MARK, UNMARK, TODO, DEADLINE, EVENT, DELETE, FIND
+    }
```
I like that you included enum into your code 👍

1001mei replied:
I agree with @atmh, using enum makes the code more clean and concise. 👍

atmh commented on src/main/java/tutu/user/Parser.java:
```diff
@@ -0,0 +1,32 @@
+        LIST, MARK, UNMARK, TODO, DEADLINE, EVENT, DELETE, FIND
+    }
+
+    public static Function checkFunction(String cmd) throws InvalidInputException {
+        if (cmd.equals("list")) {
+            return Function.LIST;
+        } else if (cmd.length() > 3 && cmd.substring(0, 4).equals("mark")) {
+            return Function.MARK;
+        } else if (cmd.length() > 5 && cmd.substring(0, 6).equals("unmark")) {
+            return Function.UNMARK;
+        } else if (cmd.length() > 3 && cmd.substring(0, 4).equals("todo")) {
+            return Function.TODO;
+        } else if (cmd.length() > 7 && cmd.substring(0, 8).equals("deadline")) {
+            return Function.DEADLINE;
+        } else if (cmd.length() > 4 && cmd.substring(0, 5).equals("event")) {
+            return Function.EVENT;
+        } else if (cmd.length() > 5 && cmd.substring(0, 6).equals("delete")) {
+            return Function.DELETE;
+        } else if (cmd.length() > 4 && cmd.substring(0, 4).equals("find")) {
+            return Function.FIND;
+        } else {
+            throw new InvalidInputException("☹ OOPS!!! I'm sorry, but I don't know what that means :-(");
+        }
```
Perhaps a switch statement would be neater. You could also consider splitting instead of substring. E.g.
```java
String[] splitSpace = input.split(" ");
String cmd = splitSpace[0];

switch (cmd) {
case "mark":
    ...
    break;
default:
    ...
}
```

atmh commented on src/main/java/tutu/Tutu.java:
```diff
@@ -0,0 +1,90 @@
+                        if (i <= taskList.taskNumber()) {
+                            taskList.mark(i);
+                            store.update(taskList);
+                        } else {
+                            System.out.println("Oops! Index is out of range");
+                        }
+                        break;
+                    case UNMARK:
+                        int j = Integer.parseInt(cmd.substring(7)); // throw NumberFormatException if not valid
+                        if (j <= taskList.taskNumber()) {
+                            taskList.unmark(j);
+                            store.update(taskList);
+                        } else {
+                            System.out.println("Oops! Index is out of range");
+                        }
+                        break;
+                    case TODO:
+                        ToDo t = new ToDo(cmd);
+                        taskList.add(t);
+                        store.update(taskList);
+                        break;
+                    case DEADLINE:
+                        Deadline dl = new Deadline(cmd);
+                        taskList.add(dl);
+                        store.update(taskList);
+                        break;
+                    case EVENT:
+                        Event e = new Event(cmd);
+                        taskList.add(e);
+                        store.update(taskList);
+                        break;
+                    case DELETE:
+                        taskList.delete(cmd);
+                        store.update(taskList);
+                        break;
+                    case FIND:
+                        taskList.find(cmd);
+                        break;
+                    default:
+                        throw new InvalidInputException("☹ OOPS!!! I'm sorry, but I don't know what that means :-(");
```
For the case and default clauses in switch statements, there should be no indentations. E.g.
```java
switch (condition) {
case ABC:
    statements;
    break;
default:
    statements;
    break;
}
```

atmh reviewed (commented):
Looks good 😄. 
Just some minor changes to be made regarding the coding standard 👍.

1001mei commented on src/main/java/tutu/data/TaskList.java:
```diff
@@ -0,0 +1,116 @@
+        return items.size();
+    }
+
+    /**
+     * Retrieves the task at a specified index.
+     * @param i Index of task to be retrieved.
+     * @return Task object representing the task at index i.
+     */
```
Perhaps adding an empty line between description and parameter section is better according to Java coding standard? I noticed this issue in several other places too.
```suggestion
    /**
     * Retrieves the task at a specified index.
     *
     * @param i Index of task to be retrieved.
     * @return Task object representing the task at index i.
     */
```

1001mei commented on src/main/java/tutu/data/TaskList.java:
```diff
@@ -0,0 +1,116 @@
+     * Returns the number of tasks in the TaskList object.
+     * @return Number of tasks.
+     */
+    public int taskNumber() {
```
Method name would be better to start with verbs. For example, consider name such as `getNumberOfTasks`. What do you think?

1001mei commented on src/main/java/tutu/task/ToDo.java:
```diff
@@ -0,0 +1,27 @@
+     */
+    public ToDo(String task) throws InvalidInputException {
+        super(task);
+        if (super.task.length() < 6) {
```
I feel like using 6 as a magic number could be avoided. Perhaps you can consider introducing a constant variable like `private final int MINIMUM_DESCRIPTION_LENGTH = 6`?

1001mei commented on src/main/java/tutu/task/Task.java:
```diff
@@ -0,0 +1,49 @@
+/** Represents a task. */
+public class Task {
+    /** Description of task */
+    protected String task;
```
I guess using name like `taskDescription` here would be more precise? Maybe it's just me, but `task` somehow gives me a feel that it is an instance of class `Task`.

1001mei reviewed (commented):
Overall, I found your code easy to read and I am able to see the class structures in it without much effort. Keep up the good work! There are some coding standard violations though, maybe you can consider using [checkstyle](https://se-education.org/guides/tutorials/checkstyle.html) to help you resolve them.

louisdavinlie commented on src/main/java/tutu/Tutu.java:
```diff
@@ -0,0 +1,90 @@
+    /**
+     * Prints a line of asterisks to separate between user inputs and resulting outputs.
+     */
+    public static void separator() {
```
According to the Java coding standard, names representing methods must be verbs. Perhaps you can add a verb to the prefix of your current method name such as print.
```suggestion
    public static void printSeparator() {
```

louisdavinlie commented on src/main/java/tutu/task/Deadline.java:
```diff
@@ -0,0 +1,48 @@
+     * @return Description of task and deadline.
+     */
+    @Override
+    public String isDone() {
```
Do you think it would be better if the name of this method is toString? I feel that it would be more proper since this method returns the String representation of the task. I think isDone sounds more like a name for Boolean variables/methods.
```suggestion
    public String toString() {
```

louisdavinlie commented on src/main/java/tutu/task/Task.java:
```diff
@@ -0,0 +1,49 @@
+    /** Description of task */
+    protected String task;
+    /** Boolean to represent if the task is done */
+    protected boolean done;
```
Consider adding "is" before the variable name to make it sound more like a Boolean variable, according to the Java coding standard.
```suggestion
    protected boolean isDone;
```

louisdavinlie reviewed (commented):
Looks excellent! The code in general is easy to read and follow

tanchinkiat99 committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

tanchinkiat99 committed:
Add checkstyle configuration files to detect coding standard violations

tanchinkiat99 committed:
Added basic GUI

tanchinkiat99 committed:
Refactor code into different classes for FXML implementation

tanchinkiat99 committed:
Integrated GUI with main application

tanchinkiat99 committed:
Improve code quality

tanchinkiat99 committed:
Add assert statements

There are several methods that require a valid integer input or integer from a String input to refer to a task in the task list.

Using assert statements in those functions helps to detect errors within the programme which can cause these integers to be invalid.

tanchinkiat99 committed:
Tutu class: ensure getResponse function follows Single Level of Abstraction Principle

MARK and UNMARK cases had different levels of abstraction than other cases.

Abstract checking of whether index input is within range to mark and unmark functions.

This is better as it ensures that level of abstraction will be the same as other cases.

tanchinkiat99 committed:
Task and its subclasses: reduce use of "magic numbers"

Declared static final variables for minimum input lengths and empty input response to increase ease of readability of code.

tanchinkiat99 committed:
TaskList class: reduce use of "magic numbers"

Declared static final variables for minimum input lengths and empty input response to increase ease of readability of code.

tanchinkiat99 committed:
Deadline and Event class: reduce use of "magic numbers"

Declared static final variables for minimum input lengths and empty input response to increase ease of readability of code.

Also used guard conditions to emphasise "happy path" to improve readability.

tanchinkiat99 committed:
Merge pull request #3 from tanchinkiat99/branch-A-Assertions

Add assert statements

tanchinkiat99 committed:
Resolve merge with master

tanchinkiat99 committed:
Merge pull request #4 from tanchinkiat99/branch-A-CodeQuality

Branch a code quality

tanchinkiat99 committed:
Debug issues with gradle

tanchinkiat99 committed:
Add basic schedule functionality.

User can now view schedule for a specific day using the command format "schedule for yyyy-dd-mm".

tanchinkiat99 committed:
Add missing javadoc comments

tanchinkiat99 committed:
Merge pull request #5 from tanchinkiat99/branch-B-ViewSchedule

Branch b view schedule

tanchinkiat99 committed:
Improve styling to make interface more pleasing to users' eyes

tanchinkiat99 committed:
Add error alert messages to draw user's attention when input is problematic

tanchinkiat99 committed:
Add exit confirmation message

tanchinkiat99 committed:
Add exit confirmation message

tanchinkiat99 committed:
Refine styling

tanchinkiat99 committed:
Correct code style

tanchinkiat99 committed:
Add window title

tanchinkiat99 committed:
Merge pull request #6 from tanchinkiat99/branch-A-BetterGui

Branch a better gui

tanchinkiat99 committed:
Add user guide

tanchinkiat99 committed:
Merge pull request #7 from tanchinkiat99/branch-A-BetterGui

Add user guide

tanchinkiat99 committed:
Minor adjustment

tanchinkiat99 committed:
Merge pull request #8 from tanchinkiat99/branch-A-UserGuide

Minor adjustment

tanchinkiat99 committed:
Update README.md

tanchinkiat99 committed:
Merge pull request #9 from tanchinkiat99/branch-A-UserGuide

Update README.md

tanchinkiat99 committed:
Set theme jekyll-theme-hacker

tanchinkiat99 committed:
Update README.md

tanchinkiat99 committed:
Merge pull request #10 from tanchinkiat99/branch-A-UserGuide

Update README.md
